README.md
# IANA Language Tags for Python #

[![Pypi package](https://badge.fury.io/py/language_tags.png)](http://badge.fury.io/py/language_tags)
[![Build Status](https://travis-ci.org/OnroerendErfgoed/language-tags.png?branch=master)](https://travis-ci.org/OnroerendErfgoed/language-tags)
[![Coverage Status](https://coveralls.io/repos/OnroerendErfgoed/language-tags/badge.png)](https://coveralls.io/r/OnroerendErfgoed/language-tags)
[![Documentation Status](https://readthedocs.org/projects/language-tags/badge/?version=latest)](https://readthedocs.org/projects/language-tags/?badge=latest)

Based on [BCP 47](http://tools.ietf.org/html/bcp47) ([RFC 5646](http://tools.ietf.org/html/rfc5646)) and the latest [IANA language subtag registry](http://www.iana.org/assignments/language-subtag-registry).

This project will be updated as the standards change.

## JSON data ##

See the [language-subtag-registry](https://github.com/mattcg/language-subtag-registry) project for the underlying JSON data.

## Javascript version ##

This project is a Python version of the [language-tags](https://github.com/mattcg/language-tags) Javascript project.


HISTORY.md
# 0.1.0 #

* Initial version

tests/test_Tag.py
# -*- coding: utf-8 -*-
import unittest
from language_tags.Tag import Tag


class TestTag(unittest.TestCase):

    def test_type(self):
        self.assertIn(Tag("en-GB-oed").type, 'grandfathered')
        self.assertIn(Tag("az-Arab").type, 'redundant')
        self.assertIn(Tag("uz-Cyrl").type, 'redundant')
        self.assertIn(Tag("zh-cmn-Hant").type, 'redundant')
        self.assertIn(Tag("mt").type, 'tag')

    def test_subtags_correct_type(self):
        tag = Tag('en')
        subtags = tag.subtags
        self.assertEqual(len(subtags), 1)
        self.assertEqual(subtags[0].type, 'language')
        self.assertEqual(subtags[0].format, 'en')

        # Lowercase - lookup should be case insensitive.
        tag = Tag('en-mt')
        subtags = tag.subtags
        self.assertEqual(len(subtags), 2)
        self.assertEqual(subtags[0].type, 'language')
        self.assertEqual(subtags[0].format, 'en')
        self.assertEqual(subtags[1].type, 'region')
        self.assertEqual(subtags[1].format, 'MT')

        tag = Tag('en-mt-arab')
        subtags = tag.subtags
        self.assertEqual(len(subtags), 3)
        self.assertEqual(subtags[0].type, 'language')
        self.assertEqual(subtags[0].format, 'en')
        self.assertEqual(subtags[1].type, 'region')
        self.assertEqual(subtags[1].format, 'MT')
        self.assertEqual(subtags[2].type, 'script')
        self.assertEqual(subtags[2].format, 'Arab')

    def test_only_existent_subtags(self):
        tag = Tag('hello')
        self.assertEqual(tag.subtags, [])

        tag = Tag('en-hello')
        subtags = tag.subtags
        self.assertEqual(len(subtags), 1)
        self.assertEqual(subtags[0].type, 'language')
        self.assertEqual(subtags[0].format, 'en')

    def test_subtags_of_private_tags(self):
        tag = Tag('en-GB-x-Beano')
        subtags = tag.subtags
        self.assertEqual(len(subtags), 2)
        self.assertEqual(subtags[0].type, 'language')
        self.assertEqual(subtags[0].format, 'en')
        self.assertEqual(subtags[1].type, 'region')
        self.assertEqual(subtags[1].format, 'GB')

    def test_subtags_of_grandfathered_tag(self):
        tag = Tag('en-GB-oed')
        self.assertEqual(tag.type, 'grandfathered')
        subtags = tag.subtags
        self.assertEqual(subtags, [])

    def test_subtags_of_redundant_tag(self):
        tag = Tag('az-Arab')
        self.assertEqual(tag.type, 'redundant')
        subtags = tag.subtags
        self.assertEqual(subtags, [])

    def test_errors_deprecated_grandfathered(self):
        # Grandfathered and deprecated, therefore invalid.
        tag = Tag('art-lojban')
        self.assertEqual(tag.type, 'grandfathered')
        self.assertIsNotNone(tag.deprecated)
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_DEPRECATED)
        self.assertEqual(err.tag, 'art-lojban')
        self.assertEqual(err.message, 'The tag art-lojban is deprecated. Use \'jbo\' instead.')
        self.assertIn(str(err.code), err.__str__())
        self.assertIn(err.message, err.__str__())
        self.assertIn(err.tag, err.__str__())
        self.assertIn(str(err.subtag), err.__str__())

    def test_errors_deprecated_redundant(self):
        # Redundant and deprecated, therefore invalid.
        tag = Tag('zh-cmn')
        self.assertEqual(tag.type, 'redundant')
        self.assertIsNotNone(tag.deprecated)
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_DEPRECATED)
        self.assertEqual(err.tag, 'zh-cmn')
        self.assertEqual(err.message, 'The tag zh-cmn is deprecated. Use \'cmn\' instead.')

    def test_errors_deprecated_subtags(self):
        # Moldovan (mo) is deprecated as a language.
        tag = Tag('mo')
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_SUBTAG_DEPRECATED)
        self.assertEqual(err.message, 'The subtag \'mo\' is deprecated.')

        # Neutral Zone (NT) is deprecated as a region.
        tag = Tag('en-NT')
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_SUBTAG_DEPRECATED)
        self.assertEqual(err.message, 'The subtag \'NT\' is deprecated.')

    def test_empty_errors_if_valid(self):
        self.assertEqual(len(Tag('en').errors), 0)

    def test_errors_no_language_tag(self):
        # test with empty tag.
        tag = Tag('')
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_NO_LANGUAGE)
        self.assertEqual(err.message, 'Empty tag.')

        tag = Tag('IQ-Arab')
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_NO_LANGUAGE)
        self.assertEqual(err.message, 'Missing language tag in \'iq-arab\'.')

        tag = Tag('419')
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_NO_LANGUAGE)
        self.assertEqual(err.message, 'Missing language tag in \'419\'.')

    def test_errors_front_no_language_tag(self):
        tag = Tag('GB-en')
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_NO_LANGUAGE)
        self.assertEqual(err.message, 'Missing language tag in \'gb-en\'.')

    def test_errors_multiple_language_subtags(self):
        tag = Tag('en-en')
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_EXTRA_LANGUAGE)
        self.assertEqual(err.message, 'Extra language subtag \'en\' found.')

        tag = Tag('en-en-GB')
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_EXTRA_LANGUAGE)
        self.assertEqual(err.message, 'Extra language subtag \'en\' found.')

        tag = Tag('ko-en')
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_EXTRA_LANGUAGE)
        self.assertEqual(err.message, 'Extra language subtag \'en\' found.')

    def test_errors_multiple_region_subtags(self):
        tag = Tag('en-GB-GB')
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_EXTRA_REGION)
        self.assertEqual(err.message, 'Extra region subtag \'GB\' found.')

        tag = Tag('ko-mt-mt')
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_EXTRA_REGION)
        self.assertEqual(err.message, 'Extra region subtag \'MT\' found.')

    def test_errors_multiple_script_subtags(self):
        tag = Tag('mt-Arab-Arab')
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_EXTRA_SCRIPT)
        self.assertEqual(err.message, 'Extra script subtag \'Arab\' found.')

        tag = Tag('en-Cyrl-Latn')
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_EXTRA_SCRIPT)
        self.assertEqual(err.message, 'Extra script subtag \'Latn\' found.')

        # First error should be regarding suppress-script, second should be regarding extra script.
        tag = Tag('en-Latn-Cyrl')
        errs = tag.errors
        self.assertEqual(len(errs), 2)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_SUPPRESS_SCRIPT)
        self.assertEqual(err.message, 'The script subtag \'Latn\' is the same as the language suppress-script.')
        err = errs[1]
        self.assertEqual(err.code, tag.ERR_EXTRA_SCRIPT)
        self.assertEqual(err.message, 'Extra script subtag \'Cyrl\' found.')

    def test_errors_multiple_extlang_subtags(self):
        tag = Tag('en-asp-bog')
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_EXTRA_EXTLANG)
        self.assertEqual(err.message, 'Extra extlang subtag \'bog\' found.')

    def test_errors_duplicate_variant_subtags(self):
        tag = Tag('ca-valencia-valencia')
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_DUPLICATE_VARIANT)
        self.assertEqual(err.message, 'Duplicate variant subtag \'valencia\' found.')

    def test_errors_length_private_use_subtags(self):
        tag = Tag('en-x-more-than-eight-chars')
        errs = tag.errors
        self.assertEqual(len(errs), 0)

        tag = Tag('en-x-morethaneightchars')
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_TOO_LONG)
        self.assertEqual(err.message, 'The private-use subtag \'morethaneightchars\' is too long.')

    def test_errors_supress_script(self):
        tag = Tag('gsw-Latn')
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_SUPPRESS_SCRIPT)
        self.assertEqual(err.message, 'The script subtag \'Latn\' is the same as the language suppress-script.')

        tag = Tag('en-Latn-GB')
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_SUPPRESS_SCRIPT)
        self.assertEqual(err.message, 'The script subtag \'Latn\' is the same as the language suppress-script.')

    def test_errors_wrong_order_subtags(self):
        tag = Tag('mt-MT-Arab')
        errs = tag.errors
        self.assertEqual(len(errs), 1)
        err = errs[0]
        self.assertEqual(err.code, tag.ERR_WRONG_ORDER)
        self.assertEqual(err.message, 'The subtag \'MT\' should not appear before \'Arab\'.')

    def test_valid_tag(self):
        self.assertTrue(Tag('en').valid)
        self.assertTrue(Tag('en-GB').valid)
        self.assertTrue(Tag('gsw').valid)
        self.assertTrue(Tag('de-CH').valid)

        # returns true for subtag followed by private tag
        self.assertTrue(Tag('en-x-whatever').valid)

    def test_valid_tag_grandfathered(self):
        # Grandfathered but not deprecated, therefore valid.
        tag = Tag('en-GB-oed')
        self.assertEqual(tag.type, 'grandfathered')
        self.assertIsNone(tag.deprecated)
        self.assertTrue(tag.valid)

    def test_valid_tag_redundant(self):
        # Redundant but not deprecated, therefore valid.
        tag = Tag('zh-Hans')
        self.assertEqual(tag.type, 'redundant')
        self.assertIsNone(tag.deprecated)
        self.assertTrue(tag.valid)

        tag = Tag('es-419')
        self.assertEqual(tag.type, 'redundant')
        self.assertIsNone(tag.deprecated)
        self.assertTrue(tag.valid)

    def test_non_valid_non_existent(self):
        self.assertFalse(Tag('zzz').valid)
        self.assertFalse(Tag('zzz-Latn').valid)
        self.assertFalse(Tag('zzz-Lzz').valid)

    def test_non_valid_deprecated_grandfathered(self):
        tag = Tag('art-lojban')
        self.assertEqual(tag.type, 'grandfathered')
        self.assertIsNotNone(tag.deprecated)
        self.assertFalse(tag.valid)

    def test_non_valid_deprecated_redundant(self):
        # Redundant and deprecated, therefore invalid.
        tag = Tag('zh-cmn')
        self.assertEqual(tag.type, 'redundant')
        self.assertIsNotNone(tag.deprecated)
        self.assertFalse(tag.valid)

        tag = Tag('zh-cmn-Hans')
        self.assertEqual(tag.type, 'redundant')
        self.assertIsNotNone(tag.deprecated)
        self.assertFalse(tag.valid)

    def test_non_valid_deprecated_subtags(self):
        #  Moldovan (mo) is deprecated as a language.
        self.assertFalse(Tag('mo').valid)

        #  Neutral Zone (NT) is deprecated as a region.
        self.assertFalse(Tag('en-NT').valid)

    def test_non_valid_redundant_script(self):
        # Swiss German (gsw) has a suppress script of Latn.
        self.assertFalse(Tag('gsw-Latn').valid)

    def test_non_valid_no_language(self):
        self.assertFalse(Tag('IQ-Arab').valid)
        self.assertFalse(Tag('419').valid)

    def test_non_valid_no_front_language(self):
        self.assertFalse(Tag('GB-en').valid)

    def test_non_valid_multiple_languages(self):
        self.assertFalse(Tag('en-en').valid)
        self.assertFalse(Tag('ko-en').valid)

    def test_non_valid_multiple_regions(self):
        self.assertFalse(Tag('en-001-gb').valid)
        self.assertFalse(Tag('gb-001').valid)

    def test_non_valid_multiple_extlangs(self):
        self.assertFalse(Tag('en-asp-bog').valid)

    def test_non_valid_multiple_scripts(self):
        self.assertFalse(Tag('arb-Latn-Cyrl').valid)

    def test_non_valid_duplicate_variant(self):
        self.assertFalse(Tag('ca-valencia-valencia').valid)

    def test_non_valid_more_private_use_lenght(self):
        self.assertFalse(Tag('en-x-morethaneightchars').valid)

    def test_non_valid_suppress_script(self):
        self.assertFalse(Tag('en-Latn').valid)
        self.assertFalse(Tag('en-GB-Latn').valid)
        self.assertFalse(Tag('gsw-Latn').valid)

    def test_deprecated(self):
        # Redundant and deprecated.
        tag = Tag('zh-cmn-Hant')
        self.assertEqual(tag.type, 'redundant')
        self.assertEqual(tag.deprecated, '2009-07-29')

        # Redundant but not deprecated.
        tag = Tag('zh-Hans')
        self.assertEqual(tag.type, 'redundant')
        self.assertIsNone(tag.deprecated)

        # Grandfathered and deprecated.
        tag = Tag('zh-xiang')
        self.assertEqual(tag.type, 'grandfathered')
        self.assertEqual(tag.deprecated, '2009-07-29')

        # Grandfathered but not deprecated.
        tag = Tag('en-GB-oed')
        self.assertEqual(tag.type, 'grandfathered')
        self.assertIsNone(tag.deprecated)

        self.assertIsNone(Tag('en').deprecated)

    def test_added(self):
        # Redundant and deprecated.
        tag = Tag('zh-cmn-Hant')
        self.assertEqual(tag.type, 'redundant')
        self.assertEqual(tag.added, '2005-07-15')

        # Redundant but not deprecated.
        tag = Tag('zh-Hans')
        self.assertEqual(tag.type, 'redundant')
        self.assertIsNone(tag.deprecated)
        self.assertEqual(tag.added, '2003-05-30')

        # Grandfathered and deprecated.
        tag = Tag('zh-xiang')
        self.assertEqual(tag.type, 'grandfathered')
        self.assertEqual(tag.added, '1999-12-18')

        # Grandfathered but not deprecated.
        tag = Tag('en-GB-oed')
        self.assertEqual(tag.type, 'grandfathered')
        self.assertIsNone(tag.deprecated)
        self.assertEqual(tag.added, '2003-07-09')

        self.assertIsNone(Tag('en').added)

    def test_description(self):
        tag = Tag('en-GB-oed')
        self.assertEqual(tag.type, 'grandfathered')
        self.assertIsNone(tag.deprecated)
        self.assertEqual(tag.descriptions, ['English, Oxford English Dictionary spelling'])

        self.assertEqual(Tag('en').descriptions, [])

    def test_format(self):
        self.assertEqual(Tag('en').format, 'en')
        self.assertEqual(Tag('en-x-more-than-eight-chars').format, 'en-x-more-than-eight-chars')
        self.assertEqual(Tag('En').format, 'en')
        self.assertEqual(Tag('EN').format, 'en')
        self.assertEqual(Tag('eN').format, 'en')
        self.assertEqual(Tag('en-gb').format, 'en-GB')
        self.assertEqual(Tag('en-gb-oed').format, 'en-GB-oed')
        self.assertEqual(Tag('az-latn').format, 'az-Latn')
        self.assertEqual(Tag('ZH-hant-hK').format, 'zh-Hant-HK')

    def test_perferred(self):
        tag = Tag('zh-cmn-Hant')
        self.assertEqual(tag.type, 'redundant')
        self.assertIsNotNone(tag.deprecated)
        self.assertIsNotNone(tag.preferred)
        self.assertEqual(tag.preferred.format, 'cmn-Hant')
        self.assertIsNone(Tag('cmn-Hant').preferred)

tests/test_tags.py
# -*- coding: utf-8 -*-
import unittest
import re

from language_tags import tags

class TestSubtag(unittest.TestCase):

    def test_get_tag(self):
        tag = tags.tag('en')
        self.assertIsNotNone(tag)

        tag = tags.tag('en-gb')
        self.assertIsNotNone(tag)
        self.assertEqual(tag.format, 'en-GB')

    def test_checks(self):
        self.assertTrue(tags.check('en'))

    def test_subtags(self):
        subtag = tags.subtags('whatever')
        self.assertEqual(subtag, [])

        subtags = tags.subtags('mt')
        self.assertEqual(len(subtags), 2)
        self.assertIn(subtags[0].type, ['language', 'region'])
        self.assertIn(subtags[0].format, ['mt', 'MT'])
        self.assertIn(subtags[1].type, ['language', 'region'])
        self.assertIn(subtags[1].format, ['mt', 'MT'])

    def test_filter(self):
        filter = tags.filter('whatever')
        self.assertIn('whatever', filter)

    def test_search(self):
        subtags = tags.search('Maltese')
        self.assertGreater(len(subtags), 0)

        self.assertEqual(subtags[0].type, 'language')
        self.assertEqual(subtags[0].format, 'mt')
        self.assertEqual(subtags[1].type, 'language')
        self.assertEqual(subtags[1].format, 'mdl')
        self.assertEqual(subtags[2].type, 'extlang')
        self.assertEqual(subtags[2].format, 'mdl')

        subtags = tags.search('Gibberish')
        self.assertEqual(subtags, [])

        subtags = tags.search('Lojban')
        self.assertEqual(len(subtags), 1)
        subtags = tags.search('Lojban', all=True)
        self.assertEqual(len(subtags), 2)

    def test_search_regexp(self):
        subtags = tags.search(re.compile('\d{4}'))
        self.assertGreater(len(subtags), 0)

    def test_type(self):
        subtag = tags.type('Latn', 'script')
        self.assertIsNotNone(subtag)
        self.assertEqual(subtag.format, 'Latn')
        self.assertEqual(subtag.type, 'script')

        self.assertIsNone(tags.type('en', 'script'))

    def test_language(self):
        subtag = tags.language('en')
        self.assertIsNotNone(subtag)
        self.assertEqual(subtag.format, 'en')
        self.assertEqual(subtag.type, 'language')

        self.assertIsNone(tags.language('GB'))

    def test_region(self):
        subtag = tags.region('IQ')
        self.assertIsNotNone(subtag)
        self.assertEqual(subtag.format, 'IQ')
        self.assertEqual(subtag.type, 'region')

    def test_languages(self):
        subtags = tags.languages('zh')
        self.assertGreater(len(subtags), 0)

        self.assertRaises(Exception, tags.languages, 'en')
        with self.assertRaises(Exception) as context:
            tags.languages('en')
        self.assertIn('\'en\' is not a macrolanguage.', context.exception.args)

    def test_date(self):
        self.assertIsNotNone(re.compile('\d{4}\-\d{2}\-\d{2}').search(tags.date()))

    def test_description(self):
        description = tags.description('nl-BE')
        self.assertIn('Dutch', description)
        self.assertIn('Flemish', description)
        self.assertIn('Belgium', description)
        description = tags.description('az-Arab')
        self.assertIn('Azerbaijani in Arabic script', description)
        description = tags.description('123')
        self.assertEqual(0, len(description))
        description = tags.description('vls')
        self.assertIn('Vlaams', description)

language_tags/Tag.py
# -*- coding: utf-8 -*-
import json
from language_tags.Subtag import Subtag
import os

parent_dir = os.path.dirname(__file__)
index = json.load(open(os.path.join(parent_dir, "data/json/index.json")))
registry = json.load(open(os.path.join(parent_dir, "data/json/registry.json")))


class Tag:
    def __init__(self, tag):
        # Lowercase for consistency (case is only a formatting convention, not a standard requirement).
        """
        Tags for Identifying Languages based on BCP 47 (RFC 5646) and the latest IANA language subtag registry.

        :param str tag: (hyphen-separated) tag.
        """
        tag = str(tag).strip().lower()

        self.data = {'tag': tag}

        if tag in index:
            types = index[tag]
            # Check if the input tag is grandfathered or redundant.
            if 'grandfathered' in types or 'redundant' in types:
                self.data['record'] = registry[types['grandfathered']] if 'grandfathered' in types \
                    else registry[types['redundant']]

        # Include errror codes
        self.ERR_DEPRECATED = 1
        self.ERR_NO_LANGUAGE = 2
        self.ERR_UNKNOWN = 3
        self.ERR_TOO_LONG = 4
        self.ERR_EXTRA_REGION = 5
        self.ERR_EXTRA_EXTLANG = 6
        self.ERR_EXTRA_SCRIPT = 7
        self.ERR_DUPLICATE_VARIANT = 8
        self.ERR_WRONG_ORDER = 9
        self.ERR_SUPPRESS_SCRIPT = 10
        self.ERR_SUBTAG_DEPRECATED = 11
        self.ERR_EXTRA_LANGUAGE = 12

    @property
    def preferred(self):
        """
        Get the preferred :class:`language_tags.Tag.Tag` of the deprecated or redundant tag.

        :return: preferred :class:`language_tags.Tag.Tag` if the deprecated or redundant tag has one, otherwise None.
        """
        if 'record' in self.data:
            return Tag(self.data['record']['Preferred-Value']) if 'Preferred-Value' in self.data['record'] else None
        else:
            return None

    @property
    def type(self):
        """
        Get the type of the tag (either grandfathered, redundant or tag see RFC 5646 section 2.2.8.).

        :return: string -- type of the tag.
        """
        if 'record' in self.data:
            return self.data['record']['Type']
        return 'tag'

    @property
    def added(self):
        """
        Get the date string of grandfathered or redundant tag when it was added to the registry.

        :return: added date string if the deprecated or redundant tag has one, otherwise None.
        """
        if 'record' in self.data:
            return self.data['record']['Added'] if 'Added' in self.data['record'] else None
        else:
            return None

    @property
    def deprecated(self):
        """
        Get the deprecation date of grandfathered or redundant tag if the tag is deprecated.

        :return: deprecation date string if the deprecated or redundant tag has one, otherwise None.
        """
        if 'record' in self.data:
            return self.data['record']['Deprecated'] if 'Deprecated' in self.data['record'] else None
        else:
            return None

    @property
    def descriptions(self):
        """
        Get the list of descriptions of the grandfathered or redundant tag.

        :return: list of descriptions. If no descriptions available, it returns an empty list.
        """
        if 'record' in self.data:
            return self.data['record']['Description'] if 'Description' in self.data['record'] else []
        else:
            return []

    @property
    def format(self):
        """
        Get format according to algorithm defined in RFC 5646 section 2.1.1.


        :return: formatted tag string.
        """
        tag = self.data['tag']
        subtags = tag.split('-')
        if len(subtags) == 1:
            return subtags[0]

        formatted_tag = subtags[0]
        private_tag = False
        for i, subtag in enumerate(subtags[1:]):

            if len(subtags[i]) == 1 or private_tag:
                formatted_tag += '-' + subtag
                private_tag = True

            elif len(subtag) == 2:
                formatted_tag += '-' + subtag.upper()

            elif len(subtag) == 4:
                formatted_tag += '-' + subtag.capitalize()
            else:
                formatted_tag += '-' + subtag

        return formatted_tag

    @property
    def subtags(self):
        """
        Get the :class:`language_tags.Subtag.Subtag` objects of the tag.

        :return: list of :class:`language_tags.Subtag.Subtag` objects that are part of the tag.
            The return list can be empty.
        """
        data = self.data
        subtags = []

        # No subtags if the tag is grandfathered or redundant.
        if 'record' in data:
            return subtags

        codes = data['tag'].split('-')
        # Try and find the language tag.
        for i, code in enumerate(codes):

            # Singletons and anything after are unhandled.
            if len(code) == 1:
                #Stop the loop (stop processing after a singleton).
                break

            # Check for non-existent tag.
            if code not in index:
                continue
            types = index[code]

            # Check against undefined because value could be 0.
            # Language subtags may only appear at the beginning of the tag, otherwise the subtag type is indeterminate.
            if 'language' in types and i == 0:
                if types['language']:
                    subtags.append(Subtag(code, 'language'))
                    continue

            if len(code) == 2:
                # Should be a region
                if 'region' in types:
                    subtags.append(Subtag(code, 'region'))
                # Error case: language subtag in the wrong place.
                elif 'language' in types:
                    subtags.append(Subtag(code, 'language'))

            elif len(code) == 3:
                # Could be a numeric region code e.g. '001' for 'World'.
                if 'region' in types:
                    subtags.append(Subtag(code, 'region'))
                elif 'extlang' in types:
                    subtags.append(Subtag(code, 'extlang'))
                # Error case: language subtag in the wrong place.
                elif 'language' in types:
                    subtags.append(Subtag(code, 'language'))

            elif len(code) == 4:
                # Could be a numeric variant
                if 'variant' in types:
                    subtags.append(Subtag(code, 'variant'))
                elif 'script' in types:
                    subtags.append(Subtag(code, 'script'))

            else:
                # Should be a variant
                if 'variant' in types:
                    subtags.append(Subtag(code, 'variant'))

        return subtags

    @property
    def valid(self):
        """
        Checks whether the tag is valid.

        :return: Bool -- True if valid otherwise False.
        """
        return len(self.errors) < 1

    @property
    def errors(self):
        """
        Get the errors of the tag.
        If invalid then the list will consist of errors containing each a code and message explaining the error.
        Each error also refers to the respective (sub)tag(s).

        :return: list of errors of the tag. If the tag is valid, it returns an empty list.
        """
        errors = []
        data = self.data
        error = self.error

        # Check if the tag is grandfathered and if the grandfathered tag is deprecated (e.g. no-nyn).
        if 'record' in data:
            if 'Deprecated' in data['record']:
                errors.append(error(self.ERR_DEPRECATED))
            # Only check every subtag if the tag is not explicitly listed as grandfathered or redundant.
            return errors

        # Check that all subtag codes are meaningful.
        codes = data['tag'].split('-')
        for i, code in enumerate(codes):
            # Ignore anything after a singleton (break)
            if len(code) < 2:
                # Check that each private-use subtag is within the maximum allowed length.
                for code in codes[i + 1:]:
                    if len(code) > 8:
                        errors.append(error(self.ERR_TOO_LONG, code))
                break

            if code not in index:
                errors.append(error(self.ERR_UNKNOWN, code))
                # Continue to the next item.
                continue

        # Check that first tag is a language tag.
        subtags = self.subtags
        if not len(subtags):
            errors.append(error(self.ERR_NO_LANGUAGE))
            return errors
        elif subtags[0].type != 'language':
            errors.append(error(self.ERR_NO_LANGUAGE))
            return errors

        # Check for more than one of some types and for deprecation.
        found = dict(language=[], extlang=[], variant=[], script=[], region=[])
        for subtag in subtags:
            type = subtag.type

            if subtag.deprecated:
                errors.append(error(self.ERR_SUBTAG_DEPRECATED, subtag))

            if type in found:
                found[type].append(subtag)

            if 'language' == type:
                if len(found['language']) > 1:
                    errors.append(error(self.ERR_EXTRA_LANGUAGE, subtag))
            elif 'region' == type:
                if len(found['region']) > 1:
                    errors.append(error(self.ERR_EXTRA_REGION, subtag))
            elif 'extlang' == type:
                if len(found['extlang']) > 1:
                    errors.append(error(self.ERR_EXTRA_EXTLANG, subtag))
            elif 'script' == type:
                if len(found['script']) > 1:
                    errors.append(error(self.ERR_EXTRA_SCRIPT, subtag))
                # Check if script is same as language suppress-script.
                else:
                    script = subtags[0].script
                    if script:
                        if script.format == subtag.format:
                            errors.append(error(self.ERR_SUPPRESS_SCRIPT, subtag))
            elif 'variant' == type:
                if len(found['variant']) > 1:
                    for variant in found['variant']:
                        if variant.format == subtag.format:
                            errors.append(error(self.ERR_DUPLICATE_VARIANT, subtag))
                            break

        # Check for correct order.
        if len(subtags) > 1:
            priority = dict(language=4, extlang=5, script=6, region=7, variant=8)
            for i, subtag in enumerate(subtags[0:len(subtags)-1]):
                next = subtags[i + 1]
                if next:
                    if priority[subtag.type] > priority[next.type]:
                        errors.append(error(self.ERR_WRONG_ORDER, [subtag, next]))

        return errors

    def error(self, code, subtag=None):
        """
        Get the :class:`language_tags.Tag.Tag.Error` of a specific Tag error code.
        The error creates a message explaining the error.
        It also refers to the respective (sub)tag(s).

        :param int code: a Tag error error:

            * 1 = Tag.ERR_DEPRECATED
            * 2 = Tag.ERR_NO_LANGUAGE
            * 3 = Tag.ERR_UNKNOWN,
            * 4 = Tag.ERR_TOO_LONG
            * 5 = Tag.ERR_EXTRA_REGION
            * 6 = Tag.ERR_EXTRA_EXTLANG
            * 7 = Tag.ERR_EXTRA_SCRIPT,
            * 8 = Tag.ERR_DUPLICATE_VARIANT
            * 9 = Tag.ERR_WRONG_ORDER
            * 10 = Tag.ERR_SUPPRESS_SCRIPT,
            * 11 = Tag.ERR_SUBTAG_DEPRECATED
            * 12 = Tag.ERR_EXTRA_LANGUAGE

        :param subtag: string (sub)tag or list of string (sub)tags creating the error.
        :return: An exception class containing: a Tag error input code, the derived message with the given (sub)tag(s).
            input
        """
        message = ""
        data = self.data

        if code == self.ERR_DEPRECATED:
            message = 'The tag %s is deprecated.' % data['tag']

            # Note that a record that contains a 'Deprecated' field and no corresponding 'Preferred-Value' field
            # has no replacement mapping (RFC 5646 section 3.1.6).
            if 'Preferred-Value' in self.data['record']:
                message += ' Use \'%s\' instead.' % data['record']['Preferred-Value']

        elif code == self.ERR_SUBTAG_DEPRECATED:
            message = 'The subtag \'%s\' is deprecated.' % subtag.format

        elif code == self.ERR_NO_LANGUAGE:
            if not len(data['tag']):
                message = 'Empty tag.'
            else:
                message = 'Missing language tag in \'%s\'.' % data['tag']

        elif code == self.ERR_UNKNOWN:
            message = 'Unknown code \'%s\'' % subtag

        elif code == self.ERR_TOO_LONG:
            message = 'The private-use subtag \'%s\' is too long.' % subtag

        elif code in [self.ERR_EXTRA_LANGUAGE,
                    self.ERR_EXTRA_EXTLANG,
                    self.ERR_EXTRA_REGION,
                    self.ERR_EXTRA_SCRIPT]:
            message = 'Extra %s subtag \'%s\' found.' % (subtag.type, subtag.format)

        elif code == self.ERR_DUPLICATE_VARIANT:
            message = 'Duplicate variant subtag \'%s\' found.' % subtag.format

        elif code == self.ERR_WRONG_ORDER:
            message = 'The subtag \'%s\' should not appear before \'%s\'.' % (subtag[0].format, subtag[1].format)

        elif code == self.ERR_SUPPRESS_SCRIPT:
            message = 'The script subtag \'%s\' is the same as the language suppress-script.' % subtag.format

        class Error(Exception):
            def __init__(self, code, message, tag, subtag):
                self.code = code
                self.message = message
                self.tag = tag
                self.subtag = subtag.format if isinstance(subtag, Subtag) else subtag

            def __str__(self):
                return repr("%s: %s (Tag %s; Subtag %s)" % (self.code, self.message, self.tag, str(self.subtag)))

        return Error(code, message, data['tag'], subtag)
